clients/src/main/java/org/apache/kafka/clients/consumer/internals/SharedConsumerState.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+
+import java.time.Duration;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class stores shared state needed by both the application thread ({@link AsyncKafkaConsumer}) and the
+ * network thread ({@link ConsumerNetworkThread}) to avoid costly inter-thread communication, where possible.
+ * This class compromises on the ideal of keeping state only in the network thread. However, this class only
+ * relies on classes which are designed to be thread-safe, thus they can be used in both the application
+ * and network threads.
+ *
+ * <p/>
+ *
+ * The following thread-safe classes are used by this class:
+ *
+ * <ul>
+ *     <li>{@link ApiVersions}</li>
+ *     <li>{@link ConsumerMetadata}</li>
+ *     <li>{@link OffsetFetcherUtils}</li>
+ *     <li>{@link SharedExceptionReference}</li>
+ *     <li>{@link SubscriptionState}</li>
+ *     <li>{@link Time}</li>
+ * </ul>
+ *
+ * <p/>
+ *
+ * In general, callers from the application thread should not mutate any of the state contained within this class.
+ * It should be considered as <em>read-only</em>, and only the network thread should mutate the state.
+ */
+public class SharedConsumerState {
+
+    private final SubscriptionState subscriptions;
+    private final OffsetFetcherUtils offsetFetcherUtils;
+    private final SharedExceptionReference updatePositionsError;
+    private final SharedExceptionReference metadataError;
+
+    public SharedConsumerState(LogContext logContext,
+                               ConsumerMetadata metadata,
+                               SubscriptionState subscriptions,
+                               Time time,
+                               long retryBackoffMs) {
+        this(logContext, metadata, subscriptions, time, retryBackoffMs, new ApiVersions());
+    }
+
+    public SharedConsumerState(LogContext logContext,
+                               ConsumerMetadata metadata,
+                               SubscriptionState subscriptions,
+                               Time time,
+                               long retryBackoffMs,
+                               ApiVersions apiVersions) {
+        requireNonNull(logContext);
+        requireNonNull(metadata);
+        requireNonNull(subscriptions);
+        requireNonNull(time);
+        requireNonNull(apiVersions);
+
+        this.subscriptions = subscriptions;
+        this.offsetFetcherUtils = new OffsetFetcherUtils(
+            logContext,
+            metadata,
+            subscriptions,
+            time,
+            retryBackoffMs,
+            apiVersions
+        );
+        this.updatePositionsError = new SharedExceptionReference();
+        this.metadataError = new SharedExceptionReference();
+    }
+
+    OffsetFetcherUtils offsetFetcherUtils() {
+        return offsetFetcherUtils;
+    }
+
+    public SharedExceptionReference updatePositionsError() {
```

Comment:
So, this sounds a bit odd because it sounds like a mutating method name. Does it update the positions error? Nope. Would positionsUpdateError be acceptable?

Reply:
You're right. I've stared at `updatePositionsError` for so long and hard that its name being confusing didn't jump out at me. I've changed it to `positionsUpdateError` here and where it's referenced by the callers.